dashboard/app/models/user.rb

```diff
@@ -605,10 +605,43 @@ def user_level_locked?(script_level, level)
     user_level.nil? || user_level.locked?(script_level.stage)
   end
 
+  # Returns the level in this script that the user most recently made progress on
+  def last_progressed_level(script)
+    user_levels.
+      where(script_id: script.id).
+      sort_by(&:updated_at).
+      last
+  end
+
+  # Returns the next script_level for the next progression level in the given
+  # script that hasn't yet been passed, starting its search at the last level we submitted
   def next_unpassed_progression_level(script)
     user_levels_by_level = user_levels_by_level(script)
 
-    script.script_levels.detect do |script_level|
+    user_level = last_progressed_level(script)
```

Comment:
In particular, can the code be refactored to define `next_unpassed`. If it is non-nil, return it (without calling `last_progressed_level`). If it is nil, compute `last_script_level` (with or without the DB query).